nano/node/websocket.cpp

```diff
@@ -0,0 +1,310 @@
+#include <algorithm>
+#include <boost/property_tree/json_parser.hpp>
+#include <chrono>
+#include <nano/node/node.hpp>
+#include <nano/node/websocket.hpp>
+
+nano::websocket::session::session (nano::websocket::listener & listener_a, boost::asio::ip::tcp::socket socket_a) :
+ws_listener (listener_a), ws (std::move (socket_a)), write_strand (ws.get_executor ())
+{
+	ws.text (true);
+}
+
+nano::websocket::session::~session ()
+{
+	ws_listener.get_node ().logger.try_log ("websocket session ended");
+}
+
+void nano::websocket::session::handshake ()
+{
+	std::lock_guard<std::mutex> lk (io_mutex);
+	ws.async_accept ([self_l = shared_from_this ()](boost::system::error_code const & ec) {
+		if (ec)
+		{
+			self_l->ws_listener.get_node ().logger.always_log ("websocket handshake failed: ", ec.message ());
+		}
+		else
+		{
+			// Start reading incoming messages
+			self_l->read ();
+		}
+	});
+}
+
+void nano::websocket::session::close ()
+{
+	std::lock_guard<std::mutex> lk (io_mutex);
+	boost::beast::websocket::close_reason reason;
+	reason.code = boost::beast::websocket::close_code::normal;
+	reason.reason = "Shutting down";
+	boost::system::error_code ec_ignore;
+	ws.close (reason, ec_ignore);
+}
+
+void nano::websocket::session::write (nano::websocket::message message_a)
+{
+	// clang-format off
+	std::lock_guard<std::mutex> lk (subscriptions_mutex);
+	if (message_a.topic == nano::websocket::topic::ack || subscriptions.find (message_a.topic) != subscriptions.end ())
+	{
+		subscriptions_mutex.unlock ();
+		boost::asio::post (write_strand,
+		[message_a, self_l = shared_from_this ()]() {
+			bool write_in_progress = !self_l->send_queue.empty ();
+			self_l->send_queue.emplace_back (message_a);
+			if (!write_in_progress)
+			{
+				self_l->write_queued_messages ();
+			}
+		});
+	}
+	// clang-format on
+}
+
+void nano::websocket::session::write_queued_messages ()
+{
+	// clang-format off
+	auto msg (send_queue.front ());
+	auto msg_str (msg.to_string ());
+
+	std::lock_guard<std::mutex> lk (io_mutex);
+	ws.async_write (boost::asio::buffer (msg_str.data (), msg_str.size ()),
+	boost::asio::bind_executor (write_strand,
+	[msg, self_l = shared_from_this ()](boost::system::error_code ec, std::size_t bytes_transferred) {
+		if (!ec)
+		{
+			self_l->send_queue.pop_front ();
```

Comment:
We need to pop_front when there's an error also right?

Reply:
Wasn't sure. If there's an intermittent error, then it'll be sent on the next try. But yeah, if keeps err'ing, the queue will fill up, so seems safest to always pop it. Changing.